In [1]:
#Import dependencies

#Import basic modules
import numpy as np
import pandas as pd
from datetime import datetime

#Import graphing modules
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()
%matplotlib inline


#Sci-Kit-Learn dependencies
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.model_selection import StratifiedShuffleSplit

#Machine Learning/Data Mining Models
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor

#Evaluation and validation
from sklearn import metrics
from sklearn.metrics import classification_report, confusion_matrix



In [2]:
#Word Development Indicators dataframe

WDI_df = pd.read_csv('WDIData.csv')
WDI_df.head()

Country Name Country Code  \
0   Arab World          ARB   
1   Arab World          ARB   
2   Arab World          ARB   
3   Arab World          ARB   
4   Arab World          ARB   

                                      Indicator Name     Indicator Code  1960  \
0  2005 PPP conversion factor, GDP (LCU per inter...      PA.NUS.PPP.05   NaN   
1  2005 PPP conversion factor, private consumptio...  PA.NUS.PRVT.PP.05   NaN   
2  Access to clean fuels and technologies for coo...     EG.CFT.ACCS.ZS   NaN   
3            Access to electricity (% of population)     EG.ELC.ACCS.ZS   NaN   
4  Access to electricity, rural (% of rural popul...  EG.ELC.ACCS.RU.ZS   NaN   

   1961  1962  1963  1964  1965  ...       2010       2011       2012  \
0   NaN   NaN   NaN   NaN   NaN  ...        NaN        NaN        NaN   
1   NaN   NaN   NaN   NaN   NaN  ...        NaN        NaN        NaN   
2   NaN   NaN   NaN   NaN   NaN  ...  82.407647  82.827636  83.169227   
3   NaN   NaN   NaN   NaN   NaN  ...  86.136134  86.782683  87.288244   
4   NaN   NaN   NaN   NaN   NaN  ...  74.543489  75.770972  76.772916   

        2013       2014       2015       2016  2017  2018  Unnamed: 63  
0        NaN        NaN        NaN        NaN   NaN   NaN          NaN  
1        NaN        NaN        NaN        NaN   NaN   NaN          NaN  
2  83.587141  83.954293  84.230630  84.570425   NaN   NaN          NaN  
3  88.389705  88.076774  88.517967  88.768654   NaN   NaN          NaN  
4  78.839139  77.487377  78.564439  78.958780   NaN   NaN          NaN  

[5 rows x 64 columns]

In [3]:
#DataFrame exploration

print(WDI_df.info())
WDI_df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 422136 entries, 0 to 422135
Data columns (total 64 columns):
Country Name      422136 non-null object
Country Code      422136 non-null object
Indicator Name    422136 non-null object
Indicator Code    422136 non-null object
1960              38296 non-null float64
1961              42017 non-null float64
1962              44198 non-null float64
1963              44084 non-null float64
1964              44611 non-null float64
1965              47305 non-null float64
1966              46902 non-null float64
1967              48633 non-null float64
1968              48209 non-null float64
1969              50011 non-null float64
1970              92892 non-null float64
1971              99126 non-null float64
1972              101898 non-null float64
1973              101523 non-null float64
1974              102838 non-null float64
1975              106739 non-null float64
1976              108656 non-null float64
1977              11198

1960          1961          1962          1963          1964  \
count  3.829600e+04  4.201700e+04  4.419800e+04  4.408400e+04  4.461100e+04   
mean   2.452211e+11  2.425950e+11  2.449366e+11  2.529916e+11  2.709654e+11   
std    1.001513e+13  1.049264e+13  1.094735e+13  1.150925e+13  1.238989e+13   
min   -3.344191e+14 -3.823000e+14 -4.327955e+14 -4.769875e+14 -5.339358e+14   
25%    4.280341e+00  4.607864e+00  4.542384e+00  4.761900e+00  4.842686e+00   
50%    3.556376e+01  3.697289e+01  3.598862e+01  3.553986e+01  3.660237e+01   
75%    2.479780e+05  2.248010e+05  1.441955e+05  1.967100e+05  2.400000e+05   
max    8.353728e+14  9.312536e+14  1.011252e+15  1.084815e+15  1.184553e+15   

               1965          1966          1967          1968          1969  \
count  4.730500e+04  4.690200e+04  4.863300e+04  4.820900e+04  5.001100e+04   
mean   2.865474e+11  3.106604e+11  3.252888e+11  3.777535e+11  4.088952e+11   
std    1.368799e+13  1.513324e+13  1.651773e+13  1.902679e+13  2.151339e+13   
min   -6.264716e+14 -7.120024e+14 -8.245134e+14 -9.379467e+14 -1.122499e+15   
25%    4.651163e+00  4.875091e+00  4.761900e+00  5.044458e+00  4.535321e+00   
50%    3.863287e+01  3.981183e+01  4.136048e+01  4.220797e+01  3.932502e+01   
75%    3.778000e+05  4.739790e+05  4.560170e+05  6.795970e+05  5.876463e+05   
max    1.388505e+15  1.544572e+15  1.724250e+15  1.979579e+15  2.302920e+15   

       ...          2010          2011          2012          2013  \
count  ...  2.358800e+05  2.263200e+05  2.277310e+05  2.229390e+05   
mean   ...  1.657079e+12  1.932391e+12  2.069415e+12  2.351690e+12   
std    ...  7.063430e+13  8.063045e+13  8.695864e+13  9.978061e+13   
min    ... -1.827709e+14 -2.168927e+14 -3.013845e+14 -3.495617e+14   
25%    ...  4.900000e+00  5.368712e+00  5.000000e+00  5.107259e+00   
50%    ...  4.245403e+01  4.711836e+01  4.629863e+01  4.664437e+01   
75%    ...  2.500000e+05  5.396312e+05  4.566098e+05  6.809700e+05   
max    ...  6.864133e+15  7.831726e+15  8.676844e+15  1.000180e+16   

               2014          2015          2016          2017          2018  \
count  2.277020e+05  2.217900e+05  2.129900e+05  1.866010e+05  3.024700e+04   
mean   2.499325e+12  2.668441e+12  3.025145e+12  3.572197e+12  3.223694e+10   
std    1.096046e+14  1.163139e+14  1.314594e+14  1.498426e+14  1.039081e+12   
min   -5.189385e+14 -8.530882e+14 -1.374105e+15 -1.345846e+15 -4.386237e+01   
25%    5.078215e+00  5.223753e+00  4.932370e+00  4.900000e+00  7.000000e+00   
50%    4.725001e+01  4.688376e+01  4.665330e+01  4.856844e+01  3.297873e+01   
75%    3.751433e+05  5.440110e+05  1.044740e+06  6.284045e+06  8.131200e+01   
max    1.155185e+16  1.167182e+16  1.319219e+16  1.534503e+16  6.817944e+13   

       Unnamed: 63  
count          0.0  
mean           NaN  
std            NaN  
min            NaN  
25%            NaN  
50%            NaN  
75%            NaN  
max            NaN  

[8 rows x 60 columns]

Given the structure of the data, .describe() doesn't really give relevant information

Performing a groupping  by 'Country Code' will allow to check whether there are equal amount of rows per country, in which case it is safer to assume that there are as many indicators for each of the countries.

In [4]:
WDI_df.groupby('Country Code').count()

Country Name  Indicator Name  Indicator Code  1960  1961  1962  \
Country Code                                                                   
ABW                   1599            1599            1599    87    88    90   
AFG                   1599            1599            1599   171   185   198   
AGO                   1599            1599            1599   135   142   148   
ALB                   1599            1599            1599   105   111   116   
AND                   1599            1599            1599    11    16    18   
ARB                   1599            1599            1599   138   153   159   
ARE                   1599            1599            1599   106   120   125   
ARG                   1599            1599            1599   215   228   268   
ARM                   1599            1599            1599    84    85    87   
ASM                   1599            1599            1599    23    27    27   
ATG                   1599            1599            1599   101   105   110   
AUS                   1599            1599            1599   251   289   288   
AUT                   1599            1599            1599   204   228   230   
AZE                   1599            1599            1599    84    85    87   
BDI                   1599            1599            1599   160   174   195   
BEL                   1599            1599            1599   181   189   199   
BEN                   1599            1599            1599   208   250   261   
BFA                   1599            1599            1599   184   206   238   
BGD                   1599            1599            1599   175   209   207   
BGR                   1599            1599            1599   102   114   119   
BHR                   1599            1599            1599   109   117   122   
BHS                   1599            1599            1599   124   129   145   
BIH                   1599            1599            1599    82    83    85   
BLR                   1599            1599            1599    86    85    87   
BLZ                   1599            1599            1599   133   139   147   
BMU                   1599            1599            1599    41    65    68   
BOL                   1599            1599            1599   205   237   258   
BRA                   1599            1599            1599   242   257   274   
BRB                   1599            1599            1599   107   125   132   
BRN                   1599            1599            1599   120   122   141   
...                    ...             ...             ...   ...   ...   ...   
TLA                   1599            1599            1599   182   196   212   
TLS                   1599            1599            1599    80   104   101   
TMN                   1599            1599            1599   153   173   180   
TON                   1599            1599            1599   106   113   115   
TSA                   1599            1599            1599   195   229   239   
TSS                   1599            1599            1599   152   169   176   
TTO                   1599            1599            1599   150   174   183   
TUN                   1599            1599            1599   145   158   176   
TUR                   1599            1599            1599   249   259   275   
TUV                   1599            1599            1599    13    19    19   
TZA                   1599            1599            1599   136   148   157   
UGA                   1599            1599            1599   195   204   215   
UKR                   1599            1599            1599    86    85    87   
UMC                   1599            1599            1599   184   199   203   
URY                   1599            1599            1599   222   244   252   
USA                   1599            1599            1599   253   271   288   
UZB                   1599            1599            1599    86    87    89   
VCT           

To verify this .unique() is applied

In [5]:
WDI_df.groupby('Country Code').count()['Indicator Name'].unique()

array([1599], dtype=int64)

The only value in this column is '1599', hence this is the number of indicators for each country.

In [6]:
WDI_df.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', 'Unnamed: 63'],
      dtype='object')

In [7]:
#Save a new dataframe dropping 'Country Code' and 'Unnamed: 63'

working_df = WDI_df[(WDI_df.columns[0]+','+WDI_df.columns[2]+','+','.join(WDI_df.columns[4:-1])).split(',')]

In [8]:
#Print time it takes to complete this cell of code
#https://thispointer.com/python-how-to-get-current-date-and-time-or-timestamp/

dateTimeObj = datetime.now()
print(dateTimeObj)

countries = working_df['Country Name'].unique()

data = pd.DataFrame(columns=list(working_df['Indicator Name'].unique()))


#Process the working Data Frame by transposing it by country and have the indicators become columns and have the rows by country and year

for country in countries:

    test_df = working_df[working_df['Country Name'] == country]
    test2 = test_df[test_df.columns[1:]].T

    new_columns = list(test2.iloc[0].values)
    old_columns = test2.columns
    mapping = dict(zip(old_columns, new_columns)) 

    test2 = test2.iloc[1:]
    test2.rename(columns=mapping, inplace = True)
    
    #https://www.geeksforgeeks.org/python-convert-two-lists-into-a-dictionary/
    
    mapping2 = dict(zip(list(test2.index), list(map(lambda item: country+"_"+str(item),list(test2.index))))) 
    mapping2

    test2.rename(index=mapping2, inplace = True)
    data = pd.concat([data,test2], sort = False)

    dateTimeObj = datetime.now()
    print(dateTimeObj)
    print(country)

dateTimeObj2 = datetime.now()    
print(dateTimeObj2)

2019-12-09 22:55:37.966509
2019-12-09 22:55:38.180687
Arab World
2019-12-09 22:55:38.256590
Caribbean small states
2019-12-09 22:55:38.322587
Central Europe and the Baltics
2019-12-09 22:55:38.389588
Early-demographic dividend
2019-12-09 22:55:38.447267
East Asia & Pacific
2019-12-09 22:55:38.525397
East Asia & Pacific (excluding high income)
2019-12-09 22:55:38.603528
East Asia & Pacific (IDA & IBRD countries)
2019-12-09 22:55:38.681659
Euro area
2019-12-09 22:55:38.761609
Europe & Central Asia
2019-12-09 22:55:38.838608
Europe & Central Asia (excluding high income)
2019-12-09 22:55:38.917608
Europe & Central Asia (IDA & IBRD countries)
2019-12-09 22:55:38.991055
European Union
2019-12-09 22:55:39.069154
Fragile and conflict affected situations
2019-12-09 22:55:39.169064
Heavily indebted poor countries (HIPC)
2019-12-09 22:55:39.263067
High income
2019-12-09 22:55:39.356031
IBRD only
2019-12-09 22:55:39.444492
IDA & IBRD total
2019-12-09 22:55:39.544014
IDA blend
2019-12-09 22:55:39.6

2019-12-09 22:56:51.014457
Peru
2019-12-09 22:56:51.826452
Philippines
2019-12-09 22:56:52.529783
Poland
2019-12-09 22:56:53.227199
Portugal
2019-12-09 22:56:53.934343
Puerto Rico
2019-12-09 22:56:54.637338
Qatar
2019-12-09 22:56:55.369756
Romania
2019-12-09 22:56:56.137695
Russian Federation
2019-12-09 22:56:56.900601
Rwanda
2019-12-09 22:56:57.706425
Samoa
2019-12-09 22:56:58.533427
San Marino
2019-12-09 22:56:59.356395
Sao Tome and Principe
2019-12-09 22:57:00.171439
Saudi Arabia
2019-12-09 22:57:00.930439
Senegal
2019-12-09 22:57:01.675437
Serbia
2019-12-09 22:57:02.493438
Seychelles
2019-12-09 22:57:03.642352
Sierra Leone
2019-12-09 22:57:04.457986
Singapore
2019-12-09 22:57:05.270477
Sint Maarten (Dutch part)
2019-12-09 22:57:06.059761
Slovak Republic
2019-12-09 22:57:06.841368
Slovenia
2019-12-09 22:57:07.622283
Solomon Islands
2019-12-09 22:57:08.414685
Somalia
2019-12-09 22:57:09.167469
South Africa
2019-12-09 22:57:09.965668
South Sudan
2019-12-09 22:57:10.949407
Spain
2019-1

In [9]:
data

2005 PPP conversion factor, GDP (LCU per international $)  \
Arab World_1960                                                NaN          
Arab World_1961                                                NaN          
Arab World_1962                                                NaN          
Arab World_1963                                                NaN          
Arab World_1964                                                NaN          
Arab World_1965                                                NaN          
Arab World_1966                                                NaN          
Arab World_1967                                                NaN          
Arab World_1968                                                NaN          
Arab World_1969                                                NaN          
Arab World_1970                                                NaN          
Arab World_1971                                                NaN          
Arab World_1972                                                NaN          
Arab World_1973                                                NaN          
Arab World_1974                                                NaN          
Arab World_1975                                                NaN          
Arab World_1976                                                NaN          
Arab World_1977                                                NaN          
Arab World_1978                                                NaN          
Arab World_1979                                                NaN          
Arab World_1980                                                NaN          
Arab World_1981                                                NaN          
Arab World_1982                                                NaN          
Arab World_1983                                                NaN          
Arab World_1984                                                NaN          
Arab World_1985                                                NaN          
Arab World_1986                                                NaN          
Arab World_1987                                                NaN          
Arab World_1988                                                NaN          
Arab World_1989                                                NaN          
...                                                            ...          
Zimbabwe_1989                                                  NaN          
Zimbabwe_1990                                                  NaN          
Zimbabwe_1991                                                  NaN          
Zimbabwe_1992                                                  NaN          
Zimbabwe_1993                                                  NaN          
Zimbabwe_1994                                                  NaN          
Zimbabwe_1995                                                  NaN          
Zimbabwe_1996                                                  NaN          
Zimbabwe_1997                                                  NaN          
Zimbabwe_1998                                                  NaN          
Zimbabwe_1999                                                  NaN          
Zimbabwe_2000                                                  NaN          
Zimbabwe_2001                                                  NaN          
Zimbabwe_2002                                                  NaN          
Zimbabwe_2003                                                  NaN          
Zimbabwe_2004                                                  NaN          
Zimbabwe_2005                                                  NaN          
Zimbabwe_2006                                                  NaN          
Zimbabwe_2007                                                  NaN          
Zimbabwe_2008                                                  NaN          
Zimbabwe_200

In [10]:
data.info(verbose=True, null_counts=True)

<class 'pandas.core.frame.DataFrame'>
Index: 15576 entries, Arab World_1960 to Zimbabwe_2018
Data columns (total 1599 columns):
2005 PPP conversion factor, GDP (LCU per international $)                                                                                       184 non-null object
2005 PPP conversion factor, private consumption (LCU per international $)                                                                       184 non-null object
Access to clean fuels and technologies for cooking (% of population)                                                                            3995 non-null object
Access to electricity (% of population)                                                                                                         6958 non-null object
Access to electricity, rural (% of rural population)                                                                                            6271 non-null object
Access to electricity, urban (% of urban populati

Data Frame was successfully transposed with the intended rows. This structure will be more useful for the machine learning parts of this project.

In [11]:
#Count and display the number of null values in the Data Frame
data_null = pd.DataFrame(data.isna().sum())
data_null

0
2005 PPP conversion factor, GDP (LCU per intern...  15392
2005 PPP conversion factor, private consumption...  15392
Access to clean fuels and technologies for cook...  11581
Access to electricity (% of population)              8618
Access to electricity, rural (% of rural popula...   9305
Access to electricity, urban (% of urban popula...   8617
Account ownership at a financial institution or...  15085
Account ownership at a financial institution or...  15085
Account ownership at a financial institution or...  15085
Account ownership at a financial institution or...  15085
Account ownership at a financial institution or...  15094
Account ownership at a financial institution or...  15086
Account ownership at a financial institution or...  15094
Account ownership at a financial institution or...  15086
Account ownership at a financial institution or...  15085
Adequacy of social insurance programs (% of tot...  15313
Adequacy of social protection and labor program...  15295
Adequacy of social safety net programs (% of to...  15329
Adequacy of unemployment benefits and ALMP (% o...  15495
Adjusted net enrollment rate, primary (% of pri...   9455
Adjusted net enrollment rate, primary, female (...  10072
Adjusted net enrollment rate, primary, male (% ...  10073
Adjusted net national income (annual % growth)       9459
Adjusted net national income (constant 2010 US$)     9429
Adjusted net national income (current US$)           7027
Adjusted net national income per capita (annual...   9459
Adjusted net national income per capita (consta...   9429
Adjusted net national income per capita (curren...   7030
Adjusted net savings, excluding particulate emi...   9023
Adjusted net savings, excluding particulate emi...  10662
...                                                   ...
Urban population                                      545
Urban population (% of total)                         504
Urban population growth (annual %)                    595
Urban population living in areas where elevatio...  14901
Urban poverty gap at national poverty lines (%)     15382
Urban poverty headcount ratio at national pover...  15249
Use of IMF credit (DOD, current US$)                 9165
Use of insecticide-treated bed nets (% of under...  15330
Value lost due to electrical outages (% of sale...  15288
Vitamin A supplementation coverage rate (% of c...  14289
Vulnerable employment, female (% of female empl...   9052
Vulnerable employment, male (% of male employme...   9052
Vulnerable employment, total (% of total employ...   9052
Wage and salaried workers, female (% of female ...   9052
Wage and salaried workers, male (% of male empl...   9052
Wage and salaried workers, total (% of total em...   9052
Wanted fertility rate (births per woman)            15263
Water productivity, total (constant 2010 US$ GD...  13886
Wholesale price index (2010 = 100)                  14218
Women making their own informed decisions regar...  15509
Women participating in the three decisions (own...  15436
Women who believe a husband is justified in bea...  15360
Women who believe a husband is justified in bea...  15412
Women who believe a husband is justified in bea...  15414
Women who believe a husband is justified in bea...  15411
Women who believe a husband is justified in bea...  15410
Women who believe a husband is justified in bea...  15413
Women who were first married by age 15 (% of wo...  15207
Women who were first married by age 18 (% of wo...  15119
Women's share of population ages 15+ living wit...  10990

[1599 rows x 1 columns]

In [12]:
data_null[0].max()

#At least one column (year) is completely empty.

15576

In order to apply machine learning, only those columns with <80% null values will be kept in the Data Frame

In [13]:
data_wo_null_columns = data[list(filter(lambda x: sum(data[x].isna())/len(data) < 0.8, data.columns))]
len(data_wo_null_columns.columns)

1044

There number of features went from 1599 to 1044 by eliminating these columns. Of course the <80% is an arbitrary criteria but does allow for establishing a code with a variable rule for dropping columns given a threshold.

As a next step, rows with null values will be dropped.

In [14]:
data_wo_null_columns.dropna(how = 'any', inplace = False)

Empty DataFrame
Columns: [Access to clean fuels and technologies for cooking (% of population), Access to electricity (% of population), Access to electricity, rural (% of rural population), Access to electricity, urban (% of urban population), Adjusted net enrollment rate, primary (% of primary school age children), Adjusted net enrollment rate, primary, female (% of primary school age children), Adjusted net enrollment rate, primary, male (% of primary school age children), Adjusted net national income (annual % growth), Adjusted net national income (constant 2010 US$), Adjusted net national income (current US$), Adjusted net national income per capita (annual % growth), Adjusted net national income per capita (constant 2010 US$), Adjusted net national income per capita (current US$), Adjusted net savings, excluding particulate emission damage (% of GNI), Adjusted net savings, excluding particulate emission damage (current US$), Adjusted net savings, including particulate emission damage (% of GNI), Adjusted net savings, including particulate emission damage (current US$), Adjusted savings: carbon dioxide damage (% of GNI), Adjusted savings: carbon dioxide damage (current US$), Adjusted savings: consumption of fixed capital (% of GNI), Adjusted savings: consumption of fixed capital (current US$), Adjusted savings: education expenditure (% of GNI), Adjusted savings: education expenditure (current US$), Adjusted savings: energy depletion (% of GNI), Adjusted savings: energy depletion (current US$), Adjusted savings: gross savings (% of GNI), Adjusted savings: mineral depletion (% of GNI), Adjusted savings: mineral depletion (current US$), Adjusted savings: natural resources depletion (% of GNI), Adjusted savings: net forest depletion (% of GNI), Adjusted savings: net forest depletion (current US$), Adjusted savings: net national savings (% of GNI), Adjusted savings: net national savings (current US$), Adjusted savings: particulate emission damage (% of GNI), Adjusted savings: particulate emission damage (current US$), Adolescent fertility rate (births per 1,000 women ages 15-19), Adolescents out of school (% of lower secondary school age), Adults (ages 15+) and children (ages 0-14) newly infected with HIV, Adults (ages 15+) newly infected with HIV, Age dependency ratio (% of working-age population), Age dependency ratio, old (% of working-age population), Age dependency ratio, young (% of working-age population), Agricultural land (% of land area), Agricultural land (sq. km), Agricultural machinery, tractors, Agricultural machinery, tractors per 100 sq. km of arable land, Agricultural methane emissions (% of total), Agricultural methane emissions (thousand metric tons of CO2 equivalent), Agricultural nitrous oxide emissions (% of total), Agricultural nitrous oxide emissions (thousand metric tons of CO2 equivalent), Agricultural raw materials exports (% of merchandise exports), Agricultural raw materials imports (% of merchandise imports), Agriculture, forestry, and fishing, value added (% of GDP), Agriculture, forestry, and fishing, value added (annual % growth), Agriculture, forestry, and fishing, value added (constant 2010 US$), Agriculture, forestry, and fishing, value added (constant LCU), Agriculture, forestry, and fishing, value added (current LCU), Agriculture, forestry, and fishing, value added (current US$), Agriculture, forestry, and fishing, value added per worker (constant 2010 US$), Air transport, freight (million ton-km), Air transport, passengers carried, Air transport, registered carrier departures worldwide, Alternative and nuclear energy (% of total energy use), Aquaculture production (metric tons), Arable land (% of land area), Arable land (hectares per person), Arable land (hectares), Armed forces personnel (% of total labor force), Armed forces personnel, total, Arms imports (SIPRI trend indicator values), Average grace period on new external debt commitments (years), Average grace period on new external d

This shows that in every row, there is at least one null value, even with the 'full null columns' that were previously dropped.

A new approach will be introduced by using a threshold, in which the row should have at least a given number of non-null values not to be dropped from the Data Frame.

In [15]:
threshold = 0.35
data_wo_null_columns.dropna(how = 'any', inplace = False, thresh=(1044-(threshold*1044)))

Access to clean fuels and technologies for cooking (% of population)  \
East Asia & Pacific (excluding high income)_1991                                                  NaN                     
East Asia & Pacific (excluding high income)_1992                                                  NaN                     
East Asia & Pacific (excluding high income)_1993                                                  NaN                     
East Asia & Pacific (excluding high income)_1994                                                  NaN                     
East Asia & Pacific (excluding high income)_1995                                                  NaN                     
East Asia & Pacific (excluding high income)_1996                                                  NaN                     
East Asia & Pacific (excluding high income)_1997                                                  NaN                     
East Asia & Pacific (excluding high income)_1998                                                  NaN                     
East Asia & Pacific (excluding high income)_1999                                                  NaN                     
East Asia & Pacific (excluding high income)_2000                                              39.1937                     
East Asia & Pacific (excluding high income)_2001                                              40.3488                     
East Asia & Pacific (excluding high income)_2002                                              41.2262                     
East Asia & Pacific (excluding high income)_2003                                              42.2912                     
East Asia & Pacific (excluding high income)_2004                                              43.1899                     
East Asia & Pacific (excluding high income)_2005                                               44.347                     
East Asia & Pacific (excluding high income)_2006                                              45.6916                     
East Asia & Pacific (excluding high income)_2007                                              46.8741                     
East Asia & Pacific (excluding high income)_2008                                              48.2388                     
East Asia & Pacific (excluding high income)_2009                                              49.5032                     
East Asia & Pacific (excluding high income)_2010                                              50.8409                     
East Asia & Pacific (excluding high income)_2011                                               52.032                     
East Asia & Pacific (excluding high income)_2012                                              53.2364                     
East Asia & Pacific (excluding high income)_2013                                              54.4463                     
East Asia & Pacific (excluding high income)_2014                                              55.4101                     
East Asia & Pacific (excluding high income)_2015                                              56.4242                     
East Asia & Pacific (excluding high income)_2016                                              57.3427                     
East Asia & Pacific (excluding high income)_2017                                                  NaN                     
Europe & Central Asia (excluding high income)_1992                                                NaN                     
Europe & Central Asia (excluding high income)_1993                                                NaN                     
Europe & Central Asia (excluding high income)_1994                                                NaN                     
...                                                                                               ...                     
Zimbabwe_1987                                                                                     NaN              

With a threshold of at least 35%, less than 30% of the rows are kept. This could be one approach to be used, the other one could be to leave them as is, and imputate the values by filling the null values with a given value - for instance the mean of the column. This however is not typically recommended as this might skew the data toward these values, further more there might be features in which the mean won't make sense and a domain expert will have to make the assessment on a case by case basis.

For example, if the GDP was to be averaged, there might be a situation in which there is a missing value for a country in which by its size, the GDP will not be nearly the world's average GDP. However to treat this will be cumbersome, time consuming, and will require a domain expert. So the risk is acknowledged, and the analysis will proceed accordingly.

In [16]:
trial_df = data_wo_null_columns.fillna(data_wo_null_columns.mean())
trial_df.isna().sum().unique()

array([0], dtype=int64)

There are no null values anymore.

In [17]:
data_wo_nulls = trial_df

The next step is to incorporate the Human Development Index by country and year into the DataFrame. The operation should be actually performed on the full Data Frame (data) since the HDI data only encompasses years 1990 onward, so 30 years worth of information will be lost from the Data Frame. It is a safer approach to drop the null values after this merger occurs.

Before that, some more preprocessing has to be performed on the full Data Frame.

In [18]:
indeces = list(map(lambda x: x.split('_'),data.index))
countries = list(map(lambda x: x[0], indeces))
years = list(map(lambda x: x[1], indeces))

data['year'] = years
data['countries'] = countries
data.head()

2005 PPP conversion factor, GDP (LCU per international $)  \
Arab World_1960                                                NaN          
Arab World_1961                                                NaN          
Arab World_1962                                                NaN          
Arab World_1963                                                NaN          
Arab World_1964                                                NaN          

                2005 PPP conversion factor, private consumption (LCU per international $)  \
Arab World_1960                                                NaN                          
Arab World_1961                                                NaN                          
Arab World_1962                                                NaN                          
Arab World_1963                                                NaN                          
Arab World_1964                                                NaN                          

                Access to clean fuels and technologies for cooking (% of population)  \
Arab World_1960                                                NaN                     
Arab World_1961                                                NaN                     
Arab World_1962                                                NaN                     
Arab World_1963                                                NaN                     
Arab World_1964                                                NaN                     

                Access to electricity (% of population)  \
Arab World_1960                                     NaN   
Arab World_1961                                     NaN   
Arab World_1962                                     NaN   
Arab World_1963                                     NaN   
Arab World_1964                                     NaN   

                Access to electricity, rural (% of rural population)  \
Arab World_1960                                                NaN     
Arab World_1961                                                NaN     
Arab World_1962                                                NaN     
Arab World_1963                                                NaN     
Arab World_1964                                                NaN     

                Access to electricity, urban (% of urban population)  \
Arab World_1960                                                NaN     
Arab World_1961                                                NaN     
Arab World_1962                                                NaN     
Arab World_1963                                                NaN     
Arab World_1964                                                NaN     

                Account ownership at a financial institution or with a mobile-money-service provider (% of population ages 15+)  \
Arab World_1960                                                NaN                                                                
Arab World_1961                                                NaN                                                                
Arab World_1962                                                NaN                                                                
Arab World_1963                                                NaN                                                                
Arab World_1964                                                NaN                                                                

                Account ownership at a financial institution or with a mobile-money-service provider, female (% of population ages 15+)  \
Arab World_1960                                                NaN                                                                        
Arab World_1961                                                NaN                                                                        
Arab World_1962                         

Mapped country and years as separate columns for better hadnling on indexing, mergers, etc.

Then read the HDI file to be merged with the data.

In [19]:
HDI_df = pd.read_excel('2018_all_indicators.xlsx')
HDI_df = HDI_df.loc[HDI_df['indicator_name'] == 'Human Development Index (HDI)']
HDI_df.head()

dimension  indicator_id                 indicator_name iso3  \
189  Composite indices        137506  Human Development Index (HDI)  AFG   
190  Composite indices        137506  Human Development Index (HDI)  ALB   
191  Composite indices        137506  Human Development Index (HDI)  DZA   
192  Composite indices        137506  Human Development Index (HDI)  AND   
193  Composite indices        137506  Human Development Index (HDI)  AGO   

    country_name   1990   1991   1992   1993   1994  ...   2009   2010   2011  \
189  Afghanistan    NaN    NaN    NaN    NaN    NaN  ...  0.453  0.463  0.471   
190      Albania  0.645  0.626  0.610  0.613  0.619  ...  0.729  0.741  0.752   
191      Algeria  0.577  0.581  0.587  0.591  0.595  ...  0.719  0.729  0.736   
192      Andorra    NaN    NaN    NaN    NaN    NaN  ...  0.830  0.828  0.827   
193       Angola    NaN    NaN    NaN    NaN    NaN  ...  0.522  0.520  0.534   

      2012   2013   2014   2015   2016   2017   9999  
189  0.482  0.487  0.491  0.493  0.494  0.498  0.498  
190  0.767  0.771  0.773  0.776  0.782  0.785  0.785  
191  0.740  0.745  0.747  0.749  0.752  0.754  0.754  
192  0.849  0.850  0.853  0.854  0.856  0.858  0.858  
193  0.543  0.554  0.564  0.572  0.577  0.581  0.581  

[5 rows x 34 columns]

Dropping columns: dimension, indicator_id, indicator_name (Only HDI was used), and iso3.

In [20]:
HDI_df.index = HDI_df['country_name']
HDI_df = HDI_df[HDI_df.columns[5:]]
HDI_df = HDI_df.T
HDI_df.head()

country_name  Afghanistan  Albania  Algeria  Andorra  Angola  \
1990                  NaN    0.645    0.577      NaN     NaN   
1991                  NaN    0.626    0.581      NaN     NaN   
1992                  NaN    0.610    0.587      NaN     NaN   
1993                  NaN    0.613    0.591      NaN     NaN   
1994                  NaN    0.619    0.595      NaN     NaN   

country_name  Antigua and Barbuda  Argentina  Armenia  Australia  Austria  \
1990                          NaN      0.704    0.631      0.866    0.795   
1991                          NaN      0.713    0.628      0.867    0.800   
1992                          NaN      0.720    0.580      0.868    0.805   
1993                          NaN      0.725    0.588      0.872    0.807   
1994                          NaN      0.728    0.599      0.875    0.813   

country_name  ...  United Kingdom  United States  Uruguay  Uzbekistan  \
1990          ...           0.775          0.860    0.692         NaN   
1991          ...           0.788          0.861    0.697         NaN   
1992          ...           0.801          0.867    0.700         NaN   
1993          ...           0.814          0.871    0.703         NaN   
1994          ...           0.827          0.875    0.709         NaN   

country_name  Vanuatu  Venezuela (Bolivarian Republic of)  Viet Nam  Yemen  \
1990              NaN                               0.634     0.475  0.399   
1991              NaN                               0.644     0.484  0.403   
1992              NaN                               0.651     0.496  0.405   
1993              NaN                               0.653     0.506  0.408   
1994              NaN                               0.654     0.517  0.409   

country_name  Zambia  Zimbabwe  
1990           0.401     0.491  
1991           0.402     0.494  
1992           0.403     0.481  
1993           0.409     0.476  
1994           0.408     0.473  

[5 rows x 189 columns]

In [21]:
import functools
import operator

trial_df = pd.DataFrame()

# https://stackoverflow.com/questions/952914/how-to-make-a-flat-list-out-of-list-of-lists

a = list(map(lambda x: list(map(lambda y: [x,y],HDI_df.index)), HDI_df.columns))
b = functools.reduce(operator.iconcat, a, [])
trial_df['countries'] = list(map(lambda x: x[0], b))
trial_df['year'] = list(map(lambda x: x[1], b))
trial_df = trial_df[trial_df['year'] != 9999]
trial_df['HDI'] = list(map(lambda x,y: HDI_df[x][y],trial_df['countries'],trial_df['year']))
trial_df.head()

countries  year  HDI
0  Afghanistan  1990  NaN
1  Afghanistan  1991  NaN
2  Afghanistan  1992  NaN
3  Afghanistan  1993  NaN
4  Afghanistan  1994  NaN

The trial_df has the mapped HDI indeces for every country per year from 1990 to 2017.

In [22]:
HDI_df = trial_df

In [23]:
HDI_df['year'] = HDI_df['year'].astype('int')
data['year'] = data['year'].astype('int')

In [24]:
data = data[data['year'] >= 1990]
data.head()

2005 PPP conversion factor, GDP (LCU per international $)  \
Arab World_1990                                                NaN          
Arab World_1991                                                NaN          
Arab World_1992                                                NaN          
Arab World_1993                                                NaN          
Arab World_1994                                                NaN          

                2005 PPP conversion factor, private consumption (LCU per international $)  \
Arab World_1990                                                NaN                          
Arab World_1991                                                NaN                          
Arab World_1992                                                NaN                          
Arab World_1993                                                NaN                          
Arab World_1994                                                NaN                          

                Access to clean fuels and technologies for cooking (% of population)  \
Arab World_1990                                                NaN                     
Arab World_1991                                                NaN                     
Arab World_1992                                                NaN                     
Arab World_1993                                                NaN                     
Arab World_1994                                                NaN                     

                Access to electricity (% of population)  \
Arab World_1990                                 74.3842   
Arab World_1991                                 74.3822   
Arab World_1992                                 74.3132   
Arab World_1993                                 75.3493   
Arab World_1994                                 75.7885   

                Access to electricity, rural (% of rural population)  \
Arab World_1990                                            57.7703     
Arab World_1991                                            61.4057     
Arab World_1992                                            61.9606     
Arab World_1993                                            63.0471     
Arab World_1994                                            63.5731     

                Access to electricity, urban (% of urban population)  \
Arab World_1990                                            92.2955     
Arab World_1991                                            92.2727     
Arab World_1992                                            91.5987     
Arab World_1993                                            92.4256     
Arab World_1994                                            92.6288     

                Account ownership at a financial institution or with a mobile-money-service provider (% of population ages 15+)  \
Arab World_1990                                                NaN                                                                
Arab World_1991                                                NaN                                                                
Arab World_1992                                                NaN                                                                
Arab World_1993                                                NaN                                                                
Arab World_1994                                                NaN                                                                

                Account ownership at a financial institution or with a mobile-money-service provider, female (% of population ages 15+)  \
Arab World_1990                                                NaN                                                                        
Arab World_1991                                                NaN                                                                        
Arab World_1992                         

In [25]:
temp_df1 = data[data['year'] != 2018].copy().reset_index(drop=True)
temp_df2 = HDI_df.copy().reset_index(drop=True)

In [26]:

index_list = []
for i in range(len(temp_df1)):

    if(temp_df1['countries'].iloc[i] not in temp_df2['countries'].values):
        index_list.append(i)
    
temp_df1.drop(index=index_list, inplace = True)

In [27]:

index_list2 = []
for i in range(len(temp_df2)):

    if(temp_df2['countries'].iloc[i] not in temp_df1['countries'].values):
        index_list2.append(i)

temp_df2.drop(index = index_list2, inplace=True)

In [28]:
temp_df1['HDI'] = temp_df2['HDI'].values

In [29]:
data_HDI = temp_df1.reset_index(drop=True)

In [30]:
data_HDI

2005 PPP conversion factor, GDP (LCU per international $)  \
0                                                   NaN          
1                                                   NaN          
2                                                   NaN          
3                                                   NaN          
4                                                   NaN          
5                                                   NaN          
6                                                   NaN          
7                                                   NaN          
8                                                   NaN          
9                                                   NaN          
10                                                  NaN          
11                                                  NaN          
12                                                  NaN          
13                                                  NaN          
14                                                  NaN          
15                                              15.1322          
16                                                  NaN          
17                                                  NaN          
18                                                  NaN          
19                                                  NaN          
20                                                  NaN          
21                                                  NaN          
22                                                  NaN          
23                                                  NaN          
24                                                  NaN          
25                                                  NaN          
26                                                  NaN          
27                                                  NaN          
28                                                  NaN          
29                                                  NaN          
...                                                 ...          
4534                                                NaN          
4535                                                NaN          
4536                                                NaN          
4537                                                NaN          
4538                                                NaN          
4539                                                NaN          
4540                                                NaN          
4541                                                NaN          
4542                                                NaN          
4543                                                NaN          
4544                                                NaN          
4545                                                NaN          
4546                                                NaN          
4547                                                NaN          
4548                                                NaN          
4549                                                NaN          
4550                                                NaN          
4551                                                NaN          
4552                                                NaN          
4553                                                NaN          
4554                                                NaN          
4555                                                NaN          
4556                                                NaN          
4557                                                NaN          
4558                                                NaN          
4559                                                NaN          
4560                                                NaN          
4561                                                NaN          
4562                                         

Now the same rationale to delete nulls can be applied to the Data Frame to be used for the HDI question.

In [31]:
data_HDI_wo_null_columns = data_HDI[list(filter(lambda x: sum(data_HDI[x].isna())/len(data_HDI) < 0.8, data_HDI.columns))]
len(data_HDI_wo_null_columns.columns)

1202

In [32]:
threshold = 0.35
data_HDI_wo_null_columns.dropna(how = 'any', inplace = False, thresh=(1202-(threshold*1202)))

Access to clean fuels and technologies for cooking (% of population)  \
17                                                15.99                     
18                                                17.44                     
19                                                18.84                     
20                                                20.68                     
21                                                22.33                     
22                                                24.08                     
23                                                26.17                     
24                                                27.99                     
25                                                 30.1                     
26                                                32.44                     
27                                                  NaN                     
33                                                  NaN                     
34                                                  NaN                     
35                                                  NaN                     
36                                                  NaN                     
37                                                  NaN                     
38                                                 39.8                     
39                                                 42.7                     
40                                                45.23                     
41                                                47.62                     
42                                                50.53                     
43                                                53.25                     
44                                                55.57                     
45                                                58.14                     
46                                                60.75                     
47                                                63.24                     
48                                                65.23                     
49                                                67.81                     
50                                                69.96                     
51                                                71.78                     
...                                                 ...                     
4525                                              15.08                     
4526                                              15.25                     
4527                                              15.43                     
4528                                              15.47                     
4529                                              15.71                     
4530                                               15.9                     
4531                                              15.99                     
4532                                              16.23                     
4533                                              16.31                     
4534                                              16.43                     
4535                                                NaN                     
4537                                                NaN                     
4538                                                NaN                     
4539                                                NaN                     
4540                                                NaN                     
4545                                                NaN                     
4546                                               32.4                     
4547                                              32.25                     
4548                                              32.18                     
4549                                              32.01                     
4

Keeping rows with at most 35% null values brings 4564 rows down to 2834, but the issue of null values is still there, and so it doesn't make sense to do this before filling the remaining null values with the mean.

In [ ]:
trial_hdi_df = data_HDI_wo_null_columns.fillna(data_HDI_wo_null_columns.mean())
trial_hdi_df.isna().sum().unique()

There are no null values anymore.

In [ ]:
data_HDI = trial_hdi_df

In [ ]:
data_HDI

Now we have to encode the targets (HDI). According to https://www.tutor2u.net/economics/reference/human-development-index HDI can be categorized into three different groups:

Low: 0-0.5
Medium: 0.5-0.8
High: 0.8+

There are other ways of doing this classification but this classification has the lowest cardinality (3). The lower the cardinality the better the model should respond.

In [ ]:
data_HDI['HDI'] = data_HDI['HDI'].apply(lambda x: 0 if x<= 0.5 else 1 if x<= 0.8 else 2)

In [ ]:
print(data_HDI['HDI'].unique())

Now the data is ready to be used with machine learning algorithms.

In [ ]:
print(data_HDI['HDI'].value_counts())

In [ ]:
#Percent of each value
data_HDI['HDI'].value_counts().map(lambda x: x*100/len(data_HDI['HDI']))

The data is imbalanced, and so a stratified sampling might be in order. However the more imbalanced the data is, the more biased the model might turn out, this has to be taken into account.

In [ ]:
#https://stackoverflow.com/questions/26414913/normalize-columns-of-pandas-data-frame

max_min_df = data_HDI.copy()

x = max_min_df.iloc[:,0:-3].values 
min_max_scaler = preprocessing.MinMaxScaler()
x_scaled = min_max_scaler.fit_transform(x)
max_min_df.iloc[:,0:-3] = x_scaled

In [ ]:
data_HDI = max_min_df.copy()

In [ ]:
# https://stackoverflow.com/questions/36997619/sklearn-stratified-sampling-based-on-a-column
sss = StratifiedShuffleSplit(n_splits=1, test_size=0.1, random_state=0)

X1 = data_HDI.iloc[:,1:-2]
y1 = data_HDI['HDI']

for train_index, test_index in sss.split(X1, y1):
    X1_train, X1_test = X1.iloc[train_index,:].values, X1.iloc[test_index, :].values
    y1_train, y1_test = y1[train_index], y1[test_index]

Random forest classifier has been given some initial fixed parameters (number of classifiers = 100). The rest of them will be changed by means of the for loop.

Criteria: GINI or Entropy<br>
Max depth: iterate from 1 to 15

In [ ]:
criteria = ['gini', 'entropy']
max_depth = 15

forest_df1 = pd.DataFrame(columns = ['Criterion', 'Max Depth', 'Train Score', 'Test Score'])

for j in range(len(criteria)):
    for i in range(1,max_depth+1):
        print(datetime.now())
        forest1 = RandomForestClassifier(criterion=criteria[j], random_state=0, max_depth=i, n_estimators=100).fit(X1_train, y1_train)
        score_train = forest1.score(X1_train, y1_train)
        score_test = forest1.score(X1_test, y1_test)
        forest_df1.loc[(i-1)+(j*max_depth+1)] = [criteria[j], i, score_train, score_test]
        

In [ ]:
forest_df1

In [ ]:
forest_df1.to_csv('Table1.csv', index=False)

In [ ]:
plt.style.use('seaborn-whitegrid')
fig, ax = plt.subplots(2, figsize=(12, 16))
plt.subplots_adjust(hspace = 0.3)

ax[0].plot(forest_df1['Max Depth'].loc[forest_df1['Criterion'] == 'gini'], forest_df1['Train Score'].loc[forest_df1['Criterion'] == 'gini'])
ax[0].plot(forest_df1['Max Depth'].loc[forest_df1['Criterion'] == 'gini'], forest_df1['Test Score'].loc[forest_df1['Criterion'] == 'gini'])
ax[0].set(xlim=(0, 16), ylim=(0.70, 1.01),xlabel='Max Depth', ylabel='Score Ratio',title=f'Train and Test Score vs Max Depth using Gini Index', xticks=list(range(0,31,2)))
ax[0].legend(['Training', 'Test'])

ax[1].plot(forest_df1['Max Depth'].loc[forest_df1['Criterion'] == 'entropy'], forest_df1['Train Score'].loc[forest_df1['Criterion'] == 'entropy'])
ax[1].plot(forest_df1['Max Depth'].loc[forest_df1['Criterion'] == 'entropy'], forest_df1['Test Score'].loc[forest_df1['Criterion'] == 'entropy'])
ax[1].set(xlim=(0, 16), ylim=(0.70, 1.01),xlabel='Max Depth', ylabel='Score Ratio',title=f'Train and Test Score vs Max Depth using Entropy',xticks=list(range(0,31,2)))
ax[1].legend(['Training', 'Test'])


plt.savefig('Decision Tree Scores.png')
plt.show()

In [ ]:
best_forest1 = RandomForestClassifier(criterion='entropy', random_state=0, max_depth=9, n_estimators=100).fit(X1_train, y1_train)
importance1 = best_forest1.feature_importances_
names1 = data_HDI.columns[:-3]
importance_df = pd.DataFrame(data={'Feature name': names1, 'Ratio of importance': importance1}).sort_values(ascending=False, by='Ratio of importance').reset_index(drop=True)

In [ ]:
pred1 = best_forest1.predict(X1_test)
print('Mean Absolute Error:', metrics.mean_absolute_error(y1_test, pred1))  
print('Mean Squared Error:', metrics.mean_squared_error(y1_test, pred1))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y1_test, pred1)))
print('Model Accuracy:', best_forest1.score(X1_test, y1_test))

In [ ]:
#https://stackoverflow.com/questions/39662398/scikit-learn-output-metrics-classification-report-into-csv-tab-delimited-format

class_report = pd.DataFrame(classification_report(y1_test,pred1, output_dict=True)).T
class_report.to_csv('Q1-ClasRepo.csv')
class_report

In [ ]:
mat = confusion_matrix(y1_test, pred1)
sns.heatmap(mat.T, square=True, annot=True, fmt='d', cbar=False,
            xticklabels = data_HDI['HDI'].unique(),
            yticklabels = data_HDI['HDI'].unique())
plt.xlabel('True label')
plt.ylabel('Predicted label')
plt.savefig('Q1-Conf_Matr.png')
plt.show()

In [ ]:
importance_df

In [ ]:
top_10 = importance_df.head(10)
top_10.index = range(1,11)
top_10.to_csv('Top_10_features.csv', index=True)
top_10

In [ ]:
sum(top_10['Ratio of importance'])

In [ ]:
importance_df[importance_df['Ratio of importance'] > 0.001]

In [ ]:
gini_importance = importance_df[importance_df['Ratio of importance'] > 0.001]['Feature name']
'GINI index (World Bank estimate)' in gini_importance

This is interesting, the GINI index, in other words income inequality, is not a factor at the time of deciding whether a country is developed or not, at least in more than 0.1% compared to other features. Let's see what is the ranking of it in the overall features data frame.

In [ ]:
importance_df[importance_df['Feature name'] == 'GINI index (World Bank estimate)']

This represents 0.0059%. It is a factor in deciding whether a nation is developed or not, but there are at least 976 other factors with more weight than the GINI Index.

In [ ]:
my_list = list(map(lambda i: (i,importance_df.reset_index(drop=True)['Ratio of importance'].head(i).sum()),range(100,1300,100)))
num = list(map(lambda feat: feat[0], my_list))
feat = list(map(lambda feat: feat[1], my_list))
feature_importance_df = pd.DataFrame(data = {'Number of Features': num, 'Accumulated Feature Importance': feat})
feature_importance_df.to_csv('Feature_importance.csv', index=False)
feature_importance_df

In [ ]:
feature_importance_df.plot(x='Number of Features', y = 'Accumulated Feature Importance', title='Accumulated Feature Importance vs Number of features used', figsize = (12,8));


Question 2: GINI INDEX prediction:

For this question the number of features can be reduced in order to only use those which are more relevant according to the random forest. This can help enhance the regression by paying attention to the "Curse of dimensionality".

From the Data Frame and the graph above, it can be seen that by using the most important 300 features, over 85% of the importance is accounted for. So by using these a random forest regressor can be built to predict the GINI Index without taking the HDI into account.

In [ ]:
features = importance_df['Feature name'][0:300]

In [ ]:
data_GINI = data_HDI[features].copy()
data_GINI['GINI index (World Bank estimate)'] = data_HDI['GINI index (World Bank estimate)']
data_GINI['year'] = data_HDI['year']

In [ ]:
validation_2017 = data_GINI[data_GINI['year'] == 2017]
data_GINI = data_GINI[data_GINI['year'] != 2017]


In [ ]:
y2 = data_GINI['GINI index (World Bank estimate)']
data_GINI = data_GINI.drop(columns=['GINI index (World Bank estimate)', 'year'])
data_GINI.head()

In [ ]:
'GINI index (World Bank estimate)' in data_GINI_features.columns

In [ ]:
X2 = data_GINI.values
y2 = y2.values

X2_train, X2_test, y2_train, y2_test = train_test_split(X2, y2, test_size=0.1, random_state=0)

In [ ]:
regr = RandomForestRegressor(max_depth=9, random_state=0,n_estimators=100)
regr.fit(X2_train, y2_train)

In [ ]:
pred2 = regr.predict(X2_test)
print('Mean Absolute Error:', metrics.mean_absolute_error(y2_test, pred2))  
print('Mean Squared Error:', metrics.mean_squared_error(y2_test, pred2))  
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(y2_test, pred2)))
print('Model Accuracy:', regr.score(X2_test, y2_test))

In [ ]:
print('2017 Score/Accuracy:', regr.score(validation_2017.iloc[:,:-2], validation_2017.iloc[:,-2]))